compiler.md

```diff
@@ -1,155 +1,483 @@
+
+- reduces memory and startup overhead significantly
+- AST no longer source of truth that compilers need to agree on
+- AST much simpler and smaller in size
+- TurboFan uses Ignition bytecode directly to optimize (no re-parse needed)
+- bytecode is 25-50% the size of equivalent baseline machine code
+- combines cutting-edge IR (intermediate representation) with multi-layered translation +
+  optimization pipeline
+- relaxed [sea of nodes](#sea-of-nodes) approach allows more effective reordering and optimization when generating
+  CFG
+- to achieve that fluid code motion, control flow optimizations and precise numerical range
+  analysis are used
+- clearer separation between JavaScript, V8 and the target architectures allows cleaner, more
+  robust generated code and adds flexibility
+- generates better quality machine code than Crankshaft JIT
+- crossing from JS to C++ land has been minimized using techniques like CodeStubAssembler
+- as a result optimizations can be applied in more cases and are attempted more aggressively
+- for the same reason (and due to other improvements) TurboFan inlines code more aggressively,
+  leading to even more performance improvements
+
+### Smaller Performance Cliffs
+
+- for most websites the optimizing compiler isn't important and could even hurt performance
+  (speculative optimizations aren't cheap)
+- pages need to load fast and unoptimized code needs to run fast _enough_, esp. on mobile
+  devices
+- previous V8 implementations suffered from _performance cliffs_
+  - optimized code ran super fast (focus on peak performance case)
+  - baseline performance was much lower
+  - as a result one feature in your code that caused deoptimization would affect your app's
+    performance dramatically, i.e. 100x difference
+- TurboFan improves this as
+  - widens fast path to ensure that optimized code is more flexible and can accept more types
+    of arguments
+  - reduces code memory overhead by reusing code generation parts of TurboFan to build Ignition
+    interpreter
+  - improves slow path
+
+### Startup Time Improved
+
+[watch](https://youtu.be/M1FBosB5tjM?t=43m25s)
+
+- bytecode smaller and faster to generate than machine code (crankshaft)
+- bytecode better suited for smaller icache (low end mobile)
+- code parsed + AST converted to bytecode only once and optimized from bytecode
+- data driven ICs reduced slow path cost (collected in feedback form, previously collected in code form)
+
+### Memory Usage Reduced
+
+[watch](https://youtu.be/M1FBosB5tjM?t=47m20s)
+
+- most important on mobile
+- Ignition code up to 8x smaller than Full-Codegen code (crankshaft)
+
+### Baseline Performance Improved
+
+[watch](https://youtu.be/M1FBosB5tjM?t=37m)
+
+- no longer relying on optimizing compiler for _sufficiently_ fast code
+- thus improved baseline performance allows delaying optimization until more feedback is collected
+- leads to less time and resources spent optimizing
```

Comment:
This also allows us to not optimize code that turns out to be infrequently executed.

Reply:
Addressed in fe82604